src/main/java/jenkins/plugins/mailer/tasks/MimeMessageBuilder.java

```diff
@@ -0,0 +1,264 @@
+import java.security.interfaces.RSAPublicKey;
+import java.util.Date;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class MimeMessageBuilder {
+
+    private String charset = "UTF-8";
+    private String mimeType = "text/plain";
+    private TaskListener listener;
+    private String defaultSuffix;
+    private String from;
+    private String replyTo;
+    private String subject;
+    private String body;
+    private AddressFilter recipientFilter;
+    private Set<InternetAddress> to = new LinkedHashSet<InternetAddress>();
+    private Set<InternetAddress> cc = new LinkedHashSet<InternetAddress>();
+    private Set<InternetAddress> bcc = new LinkedHashSet<InternetAddress>();
+
+    public MimeMessageBuilder() {
+        if (Jenkins.getInstance() != null) {
+            defaultSuffix = Mailer.descriptor().getDefaultSuffix();
+            from = JenkinsLocationConfiguration.get().getAdminAddress();
+            replyTo = Mailer.descriptor().getReplyToAddress();
+        }
+    }
+
+    public MimeMessageBuilder setCharset(String charset) {
+        if (StringUtils.isNotBlank(charset)) {
+            this.charset = charset;
+        }
+        return this;
+    }
+
+    public MimeMessageBuilder setMimeType(String mimeType) {
+        if (StringUtils.isNotBlank(mimeType)) {
+            this.mimeType = mimeType;
+        }
+        return this;
+    }
+
+    public MimeMessageBuilder setListener(TaskListener listener) {
+        this.listener = listener;
+        return this;
+    }
+
+    public MimeMessageBuilder setDefaultSuffix(String defaultSuffix) {
+        this.defaultSuffix = defaultSuffix;
+        return this;
+    }
+
+    public MimeMessageBuilder setFrom(String from) {
+        if (StringUtils.isNotBlank(from)) {
+            this.from = from;
+        }
+        return this;
+    }
+
+    public MimeMessageBuilder setReplyTo(String replyTo) {
+        if (StringUtils.isNotBlank(replyTo)) {
+            this.replyTo = replyTo;
+        }
+        return this;
+    }
+
+    public MimeMessageBuilder setSubject(String subject) {
+        this.subject = subject;
+        return this;
+    }
+
+    public MimeMessageBuilder setBody(String body) {
+        this.body = body;
+        return this;
+    }
+
+    public MimeMessageBuilder setRecipientFilter(AddressFilter recipientFilter) {
+        this.recipientFilter = recipientFilter;
+        return this;
+    }
+
+    public MimeMessageBuilder addRecipients(String recipients) throws UnsupportedEncodingException {
+        addRecipients(recipients, Message.RecipientType.TO);
+        return this;
+    }
+
+    public MimeMessageBuilder addRecipients(String recipients, Message.RecipientType recipientType) throws UnsupportedEncodingException {
+        if (StringUtils.isNotBlank(recipients) && recipientType != null) {
```

Comment:
Findbugs will complain....

So if you want to assure yourself that an `@Nonnull` is really non-null you would do something like

```
    public MimeMessageBuilder addRecipients(String recipients, @Nonnull Message.RecipientType recipientType) throws UnsupportedEncodingException {
        recipientType.getClass(); // throw NPE if null
        if (StringUtils.isNotBlank(recipients)) {
```

As the call to `Object.getClass()` is:

* the fastest way for the JVM to throw an NPE if null,
* not going to result in a findbugs warning of redundant null check
* a nifty trick picked up from the Java Runtime Libraries